src/rust/engine/src/rule_graph.rs

```diff
@@ -451,12 +451,20 @@ impl<'t> GraphMaker<'t> {
         // If no candidates were fulfillable, this rule is not fulfillable.
         unfulfillable_diagnostics.push(Diagnostic {
           params: params.clone(),
-          reason: format!(
-            "no rule was available to compute {} with parameter type{} {}",
-            type_constraint_str(product),
-            if params.len() > 1 { "s" } else { "" },
-            params_str(&params),
-          ),
+          reason: if params_str(&params) == "()" {
+            format!(
+              "No root rule found to compute {}. Maybe declare it as a RootRule({}).",
```

Comment:
I think that the shape of this error message should align with the other case, because in this case we failed to find any rules to satisfy the input. So maybe:
```
No rule found to compute {}. Maybe declare it as a RootRule({})?
```

Also, it would be good to align the capitalization in these error messages, but we can do that in a followup change.